runtime-complexity/memoization.md

```diff
@@ -1,154 +1,273 @@
 
-Memoization is when your code has a function that returns the same results every time given the same input. The result is stored in a cache in order to be used again without needing to take time to re-run the function. Remember that memoization is all about 
-saving time. 
+  
+
+Before we read about Memoization,we need to understand that it is a part of Dynamic Programming. **Dynamic Programming** is a method for solving a complex problem by breaking it down into a collection of simpler subproblems, solving each of those subproblems just once, and storing their solutions using a memory-based data structure (array, map,etc). Dynamic programming is mainly an optimization over plain recursion. Wherever we see a recursive solution that has repeated calls for same inputs, we can optimize it using Dynamic Programming. The idea is to simply store the results of subproblems, so that we do not have to re-comupute them when needed later. Memoization (top-down approach), along with its “sibling” tabulation (bottom-up approach), are the two main Dynamic Programming techniques.
+
+  
+
+Memoization is when your code has a function that returns the same results every time given the same input. The result is stored in a cache in order to be used again without needing to take time to re-run the function. Remember that memoization is all about
 
-Memoization and tabulation are two strategies used in dynamic programming. Memoization solves the top problem and stores the answer to be used later. Memoization is useful when you do not need to solve all the problems to get to the answer. Tabulation solves all problems starting at the bottom and stores the results in a matrix to be called on later. 
+saving time.
+
+  
+
+Memoization and tabulation are two strategies used in dynamic programming. Memoization solves the top problem and stores the answer to be used later. Memoization is useful when you do not need to solve all the problems to get to the answer. Tabulation solves all problems starting at the bottom and stores the results in a matrix to be called on later.
+
+  
 
 Take a minute to ask yourself if the bolded vocabulary on the first three slides is known or unknown. Take a minute to understand the new vocabulary by talking to a peer or searching online for examples and definitions.
 
+  
+
 On slide 5, determine if the functions are pure functions. Discuss and select the functions that
-are pure functions. On slide 6 reveal the results. State why the other functions are impure functions. 
+are pure functions. 
+
+On slide 6 reveal the results. State why the other functions are impure functions.
+
+  
+
+The Fibonacci Sequence problem is a well known example of recursion and provides a great way to see many different options for solving the problem and determining the best runtime. The Fibonacci Sequence is created so that each number is the sum of the two preceding numbers.
+
```

Comment:
Sorry, but this image doesn’t load 😕

Reply:
Now you just have a folder name. You’ll need the file name too. Test your own link in the preview to make sure it’s working as expected.